### PR TITLE
Use hdfs.Readdir() to optimize HDFS directory listings

### PR DESCRIPTION
**Pull request by blaenk**
## Description

Currently, listing directories on HDFS incurs a per-entry remote Stat() call penalty, the cost of which can really blow up on large directories with many entries (+1,000) especially when considered in addition to peripheral calls (such as validation) and the fact that minio is an intermediary to the client (whereas other clients listed below can query HDFS directly).

Because listing directories this way is expensive, the Golang HDFS library provides the [`Client.Open()`] function which creates a [`FileReader`] that is able to batch multiple calls together through the [`Readdir()`] function.

This is substantially more efficient for very large directories.

## Motivation and Context

In one case we were witnessing about +20 seconds to list a directory with 1,500 entries, admittedly large, but the Java hdfs ls utility as well as the Golang HDFS library sample ls utility were much faster.

Hadoop HDFS DFS (4.02s):

    λ ~/code/minio → use-readdir
    » time hdfs dfs -ls /directory/with/1500/entries/
    …
    hdfs dfs -ls   5.81s user 0.49s system 156% cpu 4.020 total

Golang HDFS library (0.47s):

    λ ~/code/hdfs → master
    » time ./hdfs ls -lh /directory/with/1500/entries/
    …
    ./hdfs ls -lh   0.13s user 0.14s system 56% cpu 0.478 total

mc and minio **without** optimization (16.96s):

    λ ~/code/minio → master
    » time mc ls myhdfs/directory/with/1500/entries/
    …
    ./mc ls   0.22s user 0.29s system 3% cpu 16.968 total

mc and minio **with** optimization (0.40s):

    λ ~/code/minio → use-readdir
    » time mc ls myhdfs/directory/with/1500/entries/
    …
    ./mc ls   0.13s user 0.28s system 102% cpu 0.403 total

I also received a report that on a directory of +20,000 entries, it reduced `ls` times from 4 minutes down to 1.27 seconds.

[`Client.Open()`]: https://godoc.org/github.com/colinmarc/hdfs#Client.Open
[`FileReader`]: https://godoc.org/github.com/colinmarc/hdfs#FileReader
[`Readdir()`]: https://godoc.org/github.com/colinmarc/hdfs#FileReader.Readdir

## How to test this PR?

List a large HDFS directory (mine had 1,500 entries).

You can probably create such a directory like this. Note that I include a sub-directory to test mixed contents.

``` bash
$ TEST_PATH=/some/path
$ hdfs dfs -mkdir -p $TEST_PATH/test_sub_dir
$ hdfs dfs -touchz $TEST_PATH/test_sub_dir/test_child_file
$ for i in {0001..1500}; do hdfs dfs -touchz "$TEST_PATH/file_${i}"; done
```

That might be slow since it's invoking `hdfs` per file, so you can probably batch them into one call like this:

``` bash
$ TEST_PATH=/some/path
$ hdfs dfs -mkdir -p $TEST_PATH/test_sub_dir
$ hdfs dfs -touchz $TEST_PATH/test_sub_dir/test_child_file
$ for i in {0001..1500}; do echo "$TEST_PATH/file_${i}"; done | xargs hdfs dfs -touchz
```

Then you can test by listing it:

``` bash
$ ./mc ls myhdfs$TEST_PATH
$ ./mc ls --recursive myhdfs$TEST_PATH
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by blaenk** [comment]
It seems like I need to iron out some wrinkles with `--recursive`. Looking into it.

**Comment by blaenk** [comment]
It's working now with `--recursive` listings.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#5ea2fbcb-8eb0-42c6-9517-1607a0368d60) |
<a name="5ea2fbcb-8eb0-42c6-9517-1607a0368d60"></a>
### 10121-05af42f/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31822
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 33ef4aae51f3:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 6 seconds
(5/15) Running awscli tests ... done in 1 minutes and 28 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 41 seconds
(8/15) Running minio-dotnet tests ... done in 58 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 41 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 30 seconds
(11/15) Running minio-js tests ... done in 1 minutes and 3 seconds
(12/15) Running minio-py tests ... done in 2 minutes and 35 seconds
(13/15) Running s3cmd tests ... FAILED in 23 seconds
{
  "name": "s3cmd",
  "duration": "12325",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "WARNING: Bucket is not empty. Removing all the objects from it first. This may take some time...\nERROR: S3 error: 404 (Not Found)"
}
(13/15) Running s3select tests ... done in 7 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by blaenk** [comment]
Sorry for the extra pushes, I think it should be set now, barring any review comments and/or problems found in the CI of course.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#6f9dd942-381a-4c4f-95f1-eb5c7a8711ae) |
<a name="6f9dd942-381a-4c4f-95f1-eb5c7a8711ae"></a>
### 10121-f2911ae/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31760
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e0fc50cd332b:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 22 seconds
(4/15) Running aws-sdk-ruby tests ... done in 19 seconds
(5/15) Running awscli tests ... FAILED in 53 seconds
{
  "name": "awscli",
  "duration": 1641,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 3 minutes and 28 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 49 seconds
(8/15) Running minio-go tests ... done in 5 minutes and 11 seconds
(9/15) Running minio-java tests ... done in 6 minutes and 6 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 35 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 40 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 22 seconds
(13/15) Running s3select tests ... done in 1 minutes and 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by blaenk** [comment]
Added to OP:

> I also received a report that on a directory of more than 20,000 entries, it reduced ls times from 4 minutes down to 1.27 seconds.

**Review by harshavardhana — APPROVED** [review]

File: cmd/gateway/hdfs/gateway-hdfs.go [hunk 1]
```diff
@@ -384,28 +384,82 @@ func (n *hdfsObjects) listDirFactory() minio.ListDirFunc {
...
+		filePath := n.hdfsPathJoin(filePath, fileInfo.Name())
+
+		if fileInfo.IsDir() {
+			filePath += minio.SlashSeparator
+		}
+
+		fileInfos[filePath] = fileInfo
+	}
+
+	return dirReader.Stat(), nil
```
Comment (Praveenrajmani) [inline cmd/gateway/hdfs/gateway-hdfs.go]: We can avoid this duplicate stat call as we already read the parent dir's stat info in L443

File: cmd/gateway/hdfs/gateway-hdfs.go [hunk 2]
```diff
@@ -384,28 +384,82 @@ func (n *hdfsObjects) listDirFactory() minio.ListDirFunc {
...
+
+		delete(fileInfos, filePath)
+
+		return objectInfo, nil
 	}
 
 	return minio.ListObjects(ctx, n, bucket, prefix, marker, delimiter, maxKeys, n.listPool, n.listDirFactory(), getObjectInfo, getObjectInfo)
 }
 
+func (n *hdfsObjects) populateDirectoryListing(filePath string, fileInfos map[string]os.FileInfo) (os.FileInfo, error) {
```
Comment (Praveenrajmani) [inline cmd/gateway/hdfs/gateway-hdfs.go]: Can we add a comment saying that this function will do a non-recursive/first-level listing

File: cmd/gateway/hdfs/gateway-hdfs.go [hunk 3]
```diff
@@ -384,28 +384,82 @@ func (n *hdfsObjects) listDirFactory() minio.ListDirFunc {
 
 // ListObjects lists all blobs in HDFS bucket filtered by prefix.
 func (n *hdfsObjects) ListObjects(ctx context.Context, bucket, prefix, marker, delimiter string, maxKeys int) (loi minio.ListObjectsInfo, err error) {
-	if _, err := n.clnt.Stat(n.hdfsPathJoin(bucket)); err != nil {
+	fileInfos := make(map[string]os.FileInfo)
+	directoryPath := n.hdfsPathJoin(bucket, prefix) + minio.SlashSeparator
+
+	var directoryFileInfo os.FileInfo
+
+	if directoryFileInfo, err = n.populateDirectoryListing(directoryPath, fileInfos); err != nil {
```
Comment (Praveenrajmani) [inline cmd/gateway/hdfs/gateway-hdfs.go]: getObjectInfo would still cover this prelude listing right? Why are we doing this initially and checking if the dentries are repeating (already preserved) in L404. 

Are there any performance impacts in doing this?

File: cmd/gateway/hdfs/gateway-hdfs.go (hunk 1 above)
Comment (blaenk) [reply cmd/gateway/hdfs/gateway-hdfs.go]: Good catch.

This is not an actual network/hdfs/api call, it's  [`FileReader.Stat()`](https://github.com/colinmarc/hdfs/blob/v1.1.3/file_reader.go#L55-L57) which is just a field accessor, but I can certainly reuse the return value from L443 👍🏼

File: cmd/gateway/hdfs/gateway-hdfs.go (hunk 2 above)
Comment (blaenk) [reply cmd/gateway/hdfs/gateway-hdfs.go]: Of course, will do 👍🏼

File: cmd/gateway/hdfs/gateway-hdfs.go (hunk 3 above)
Comment (blaenk) [reply cmd/gateway/hdfs/gateway-hdfs.go]: > getObjectInfo would still cover this prelude listing right? Why are we doing this initially…

Are you asking why we do `populateDirectoryListing()` here and not just leave it to be within `getObjectInfo`?

`getObjectInfo` now works by simply keying into the `path → FileInfo` map. If a path is missing, it lists its _parent's_ directory (since to get `/dir/file`'s `FileInfo` we need to list `/dir`). So in the case of the "root" path e.g. `/dir` (assuming we did `ls /dir`), it would mean listing `/`, unnecessarily populating the map with `/`'s entries when all we care about is `/dir`. This can be really bad if `/` is huge.

This process makes sense in all other cases because it would be a nested path in a `--recursive` listing.

The original removed code on line 387 performed an explicit network `Stat()` for the target `directoryPath`, which would be one way for us to "bootstrap" the root `FileInfo`, but `FileReader` is already doing this under the hood and helpfully exposes it with the `FileReader.Stat()` accessor, so, knowing that at a minimum we will want to list the first level of a directory anyway, I just do it upfront right here which allows us to preserve `directoryPath`'s `FileInfo` without an extra unnecessary call.

Not sure if I understood you correctly or if that answers your question.

> …and checking if the dentries are repeating (already preserved) in L404.

Sorry I don't follow this part, what do you mean by "checking if the dentries are repeating (already preserved)"? As far as I'm aware, there are no repeating entries.

> Are there any performance impacts in doing this?

For performing the first level's listing up-front? I don't see why. This is doing work that needs to be done either way, and subsequent `populateDirectoryListing()` will only be for things we don't already have.

So in other words:

1. Here, up-front, we populate the cache with the `FileInfo` for `directoryPath` and for its direct children.
2. On `getObjectInfo(directoryPath)`, in cache, no listing necessary
3. On `getObjectInfo(directoryPath.Child)`, in cache, no listing necessary
4. On `getObjectInfo(directoryPath.Child.Child)`, not in cache, so `list(directoryPath.Child)`

Hope that answers your questions.

File: cmd/gateway/hdfs/gateway-hdfs.go (hunk 1 above)
Comment (blaenk) [reply cmd/gateway/hdfs/gateway-hdfs.go]: I'm no longer returning this `FileInfo` so it took care of itself.

File: cmd/gateway/hdfs/gateway-hdfs.go (hunk 3 above)
Comment (Praveenrajmani) [reply cmd/gateway/hdfs/gateway-hdfs.go]: Now i see its purpose to avoid unnecessary listing :+1:

**Review by Praveenrajmani — APPROVED** [review]
Nice work @blaenk . Thanks for the PR!

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by blaenk** [comment]
Happy to help! Thank you both for working on minio.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally
